src/Packs.sol

```diff
@@ -0,0 +1,828 @@
+        address marketplace_,
+        bytes calldata orderData_,
+        uint256 orderAmount_,
+        address token_,
+        uint256 tokenId_,
+        uint256 payoutAmount_,
+        bytes calldata commitSignature_,
+        bytes calldata fulfillmentSignature_,
+        FulfillmentOption choice_,
+        bytes calldata choiceSignature_
+    ) internal nonReentrant {
+        // Basic validation of tx
+        if (marketplace_ == address(0)) revert InvalidMarketplace();
+        if (msg.value > 0) _depositTreasury(msg.value);
+        if (orderAmount_ > treasuryBalance) revert InsufficientBalance();
+        if (isFulfilled[commitId_]) revert AlreadyFulfilled();
+        if (isCancelled[commitId_]) revert CommitIsCancelled();
+        if (commitId_ >= packs.length) revert InvalidCommitId();
+        if (payoutAmount_ > orderAmount_) revert PayoutAmountGreaterThanOrderAmount();
+
+        CommitData memory commitData = packs[commitId_];
+
+        // Check the cosigner signed the commit
+        address commitCosigner = verifyCommit(commitData, commitSignature_);
+        if (commitCosigner != commitData.cosigner) revert InvalidCosigner();
+        if (!isCosigner[commitCosigner]) revert InvalidCosigner();
+
+        uint256 rng = PRNG.rng(commitSignature_);
+        bytes32 digest = hashCommit(commitData);
+        bytes32 fulfillmentHash =
+            hashFulfillment(digest, marketplace_, orderAmount_, orderData_, token_, tokenId_, payoutAmount_, choice_);
+
+        // Check the cosigner signed the order data
+        address fulfillmentCosigner = verifyHash(fulfillmentHash, fulfillmentSignature_);
+        if (fulfillmentCosigner != commitData.cosigner) revert InvalidCosigner();
+        if (!isCosigner[fulfillmentCosigner]) revert InvalidCosigner();
+
+        // Determine bucket and validate orderAmount and payoutAmount are within bucket range
+        uint256 bucketIndex = _getBucketIndex(rng, commitData.buckets);
+        BucketData memory bucket = commitData.buckets[bucketIndex];
+        if (orderAmount_ < bucket.minValue) revert InvalidOrderAmount();
+        if (orderAmount_ > bucket.maxValue) revert InvalidOrderAmount();
+        if (payoutAmount_ < bucket.minValue) revert InvalidPayoutAmount();
+        if (payoutAmount_ > bucket.maxValue) revert InvalidPayoutAmount();
+
+        // Check the fulfillment option
+        address choiceSigner = verifyHash(fulfillmentHash, choiceSignature_);
+        if (choiceSigner != commitData.receiver && choiceSigner != commitData.cosigner) revert InvalidChoiceSigner();
+
+        // If the user wants to fulfill via NFT but the option has expired, default to payout
+        FulfillmentOption fulfillmentType = choice_;
+        if (choice_ == FulfillmentOption.NFT && block.timestamp > nftFulfillmentExpiresAt[commitId_]) {
+            fulfillmentType = FulfillmentOption.Payout;
+        }
+
+        // Mark the commit as fulfilled
+        isFulfilled[commitId_] = true;
```

Comment:
Nit: Checks-Effects-Interactions. For sake of form/pattern I think we can move `isFulfilled...` up higher, around `L326` after the initial validation takes place. 

We aren't calling out to other contracts so there's no risk and I should probably have marked the PRNG immutable in earlier versions.